executor/join.go

```diff
@@ -665,46 +650,99 @@ func (e *HashJoinExec) handleFetchInnerAndBuildHashTablePanic(r interface{}) {
 }
 
 func (e *HashJoinExec) fetchInnerAndBuildHashTable(ctx context.Context) {
-	// innerResultCh transfers inner chunk from inner fetch to build hash table.
-	innerResultCh := make(chan *chunk.Chunk, 1)
-	doneCh := make(chan struct{})
-	// TODO: just used in test now, thus we return directly in this if-branch.
-	// Remained work:
-	// 1. parallel build hash tables for the inner sub-partitions
-	// 2. partition the outer relation into sub-partitions
-	// 3. parallel join corresponded inner-partitions with outer-partitions
+	// Step1: fetch all the data of the inner relation into memory.
+	if err := e.fetchInnerRows(ctx); err != nil {
+		e.innerFinished <- err
+		return
+	}
+	if e.finished.Load().(bool) {
+		return
+	}
+
+	// Step2: do parallel partition if needed.
+	needPartition := false
 	if e.ctx.GetSessionVars().EnableRadixJoin {
-		go util.WithRecovery(func() {
-			e.fetchInnerRows(ctx, innerResultCh, doneCh)
-			if !e.evalRadixBit() {
-				return
-			}
-			if e.partitionInnerRows() != nil {
+		needPartition = e.evalRadixBit()
+		if needPartition {
+			if err := e.partitionInnerRows(); err != nil {
+				e.innerFinished <- err
 				return
 			}
-
-		}, nil)
+		}
+	}
+	if e.finished.Load().(bool) {
 		return
 	}
 
-	go util.WithRecovery(func() { e.fetchInnerRows(ctx, innerResultCh, doneCh) }, nil)
+	// Step3: build hash table(s) for the inner sub-relation.
+	if err := e.buildHashTable(needPartition); err != nil {
+		e.innerFinished <- err
+	}
+}
 
-	// TODO: Parallel build hash table. Currently not support because `mvmap` is not thread-safe.
-	err := e.buildHashTableForList(innerResultCh)
-	if err != nil {
-		e.innerFinished <- errors.Trace(err)
-		close(doneCh)
+func (e *HashJoinExec) buildHashTable(hasPartitionPhase bool) error {
+	e.innerKeyColIdx = make([]int, len(e.innerKeys))
+	for i := range e.innerKeys {
+		e.innerKeyColIdx[i] = e.innerKeys[i].Index
+	}
+
+	if !hasPartitionPhase {
+		return e.buildGlobalHashTable()
 	}
-	// wait fetchInnerRows be finished.
-	for range innerResultCh {
+	return e.buildHashTable4Partitions()
+}
+
+func (e *HashJoinExec) wait4BuildHashTable(wg *sync.WaitGroup, finishedCh chan error) {
+	wg.Wait()
+	close(finishedCh)
+}
+
+func (e *HashJoinExec) buildHashTable4Partitions() error {
+	e.hashTables = make([]*mvmap.MVMap, len(e.innerParts))
+	buildFinishedCh := make(chan error, e.concurrency)
+	wg := &sync.WaitGroup{}
+	wg.Add(int(e.concurrency))
+	go e.wait4BuildHashTable(wg, buildFinishedCh)
+	for i := 0; i < int(e.concurrency); i++ {
+		workerID := i
+		go util.WithRecovery(func() {
+			defer wg.Done()
+			e.doBuild(workerID, buildFinishedCh)
+		}, nil)
 	}
+	return <-buildFinishedCh
 }
 
-// buildHashTableForList builds hash table from `list`.
+func (e *HashJoinExec) doBuild(workerID int, finishedCh chan error) {
+	var err error
+	for i := workerID; i < len(e.innerParts); i += int(e.concurrency) {
+		if e.finished.Load().(bool) {
+			return
+		}
+		if e.innerParts[i] == nil {
```

Comment:
This is an abnormal case, should we log a warning that there is an empty partition?

Reply:
This is not an abnormal case, cause `len(e.innerParts)` is the max number of partitions, but not all the slot of `e.innerParts` will be filled with rows.

Reply:
Yeah. A good hash function should map the expected inputs as evenly as possible over its output range. From the perspective of hash functions, an empty partition may be caused by:
- a bad function. I think this needs to be recorded
- a data skew happened on some join keys. I think this also needs to be recorded

Reply:
Maybe we'd better log it when `len(e.Nil_innerPart) / len(e.innerParts) > threshold(support 40%)`?

Reply:
how about logging these two things once there is at lease one empty partition in the `Close()` function:
- number of total partitions
- number of empty partitions